src/fireedge/src/server/routes/api/auth/utils.js

```diff
@@ -0,0 +1,669 @@
+         *********************************************************/
+        key: serverAdminPassword.substring(0, 32),
+        iv: serverAdminPassword.substring(0, 16),
+      })
+    }
+
+    if (tokenWithServerAdmin) {
+      // set global state
+      if (setGlobalNewToken) {
+        if (!global.users) {
+          global.users = {}
+        }
+        if (!global.users[JWTusername]) {
+          global.users[JWTusername] = { tokens: [] }
+        }
+        !validToken &&
+          global.users[JWTusername].tokens.push({
+            token: tokenWithServerAdmin.token,
+            time: parseInt(expireTime, 10),
+          })
+      }
+
+      genJWT(tokenWithServerAdmin, {
+        NAME: JWTusername,
+        ID: userData.ID,
+        TEMPLATE: userData.TEMPLATE,
+      })
+
+      next()
+    }
+  } else {
+    updaterResponse(httpResponse(internalServerError))
+    next()
+  }
+}
+
+/**
+ * Get server admin.
+ *
+ * @returns {object|undefined} data serveradmin
+ */
+const getServerAdmin = () => {
+  const serverAdminData = getSunstoneAuth()
+  const { username, key, iv } = serverAdminData
+  if (username && key && iv) {
+    return {
+      ...serverAdminData,
+      token: createTokenServerAdmin({
+        serverAdmin: username,
+        username,
+        key,
+        iv,
+      }),
+    }
+  }
+}
+
+/**
+ * Get server admin and wrap user.
+ *
+ * @param {object} userData - opennebula user data
+ */
+const getServerAdminAndWrapUser = (userData = {}) => {
+  const serverAdminData = getServerAdmin()
+  const { username, token } = serverAdminData
+  if (username && token) {
+    const oneConnect = connectOpennebula(`${username}:${username}`, token.token)
+    oneConnect({
+      action: USER_INFO,
+      parameters: getDefaultParamsOfOpennebulaCommand(USER_INFO, GET),
+      callback: (err, value) => {
+        responseOpennebula(
+          updaterResponse,
+          err,
+          value,
+          (serverAdmin = {}) => wrapUserWithServerAdmin(serverAdmin, userData),
+          next
+        )
+      },
+      fillHookResource: false,
+    })
+  }
+}
+
+/**
+ * Remote login route function.
+ *
+ * @param {string} userData - user remote data user:password
+ * @param {object} response - http response
+ */
+const remoteLogin = (userData = '', response = {}) => {
+  const serverAdminData = getServerAdmin()
+  const { username, token } = serverAdminData
+  const [usr, pss = usr] = userData.split(':')
+  if (username && token && usr && pss) {
+    const oneConnect = connectOpennebula(`${username}:${username}`, token.token)
+    oneConnect({
+      action: USER_POOL_INFO,
+      parameters: getDefaultParamsOfOpennebulaCommand(USER_POOL_INFO, GET),
+      callback: (_, value) => {
+        const users = value?.USER_POOL?.USER || []
+        if (users.length) {
+          const userFound = users.find(
+            (data) =>
+              data.NAME === usr &&
+              data.PASSWORD === pss &&
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Avoid using `==` and `===` operators with sensitive data fields like passwords, api keys, tokens etc. as they are vulnerable to timing attack. Consider using `crypto.timingSafeEqual` or `secure-compare` for time-safe string comparison.

[Learn more](https://cwe.mitre.org/data/definitions/208.html)